### DISCUSSION TITLE
200+ms TTFB for cached home page in solid-realworld

### DISCUSSION THREAD

**Discussion by ITwrx** [solidjs/solid#757] [Q&A] [answered]
Hi,

I'm evaluating a few js frameworks for my first possible attempt at learning one to use in production and i noticed a lag in the display of the articles list on the home page of the [realworld example](https://github.com/solidjs/solid-realworld) after previously visiting the page. FF shows the request as `304` cached, but still shows 200+ms TTFB, while chromium shows it as a `200` status, but with about the same TTFB (200+ms). Is it re-fetching the articles list? if so, why? I hoped all the data on a page could be cached in the browser and be served from there unless something had changed from the backend, or user input, and therefore be instant. Is this not the case? Or the realworld example chose not to implement it that way for some reason?

I also noticed that sometimes when you navigate to a specific article's page it makes requests and downloads the article and comments in around 200+ms, but sometimes it renders the page instantly with dev tools showing no request at all. Why doesn't it show it instantly like that all the time (between changes to the article/comments), and why doesn't the home page do the same thing?

Thanks

**Comment by ryansolid** [comment, marked as answer]
There have been some issues with the backend recently so I haven't updated. It looks they've resolved it so I can look at updating this example. This version is also completely client rendered(no server rendering) so there is a client-side cache when navigating around but it is gone the second you reload the page. Generally speaking one could persist to local storage but that is a whole other thing. You will find most Realworld demos refetch on page reload to my knowledge or they are server-rendered (which contains the data already). 

I host this on github pages so it is CSR only right now. If we want to change the nature of the demo would need to update hosting. I intended to do so when I completed the work on our SSR starter kit.

**Reply by ITwrx** [reply]
> There have been some issues with the backend recently so I haven't updated. It looks they've resolved it so I can look at updating this example. This version is also completely client rendered(no server rendering) so there is a client-side cache when navigating around but it is gone the second you reload the page.

so the client side cache could theoretically be used as i described?

> Generally speaking one could persist to local storage but that is a whole other thing.

Is local storage required for the "cached through reloads" functionality, or just another method/place to save the data?

> You will find most Realworld demos refetch on page reload to my knowledge or they are server-rendered (which contains the data already).

I wonder if this is due to convention, convenience, or to have a level playing field... If the caching i describe is not appropriate for the real world example then perhaps caching strategies could be in the tutorials or other Solid examples.

> I host this on github pages so it is CSR only right now. If we want to change the nature of the demo would need to update hosting. I intended to do so when I completed the work on our SSR starter kit.

Yeah, i was just focusing on the CSR aspects right now, as well. I read through many Solid issues trying to see where Solid was on the SSR stuff. When SSR comes i hope i can use it in Isomorphic mode where the first render is SSR and then SPA mode after that (like sveltekit advertises, IIRC) with the persistent CS caching described here. I will have all my DB queries cached in ram on the backend, but i don't even want the FE wasting time on the requests unless the data has changed. I saw some framework had API response caching, and i'm guessing what i describe is the same/similar. 

P.S. I'm hoping Deno can be used as the SSR engine at some point, as Node is probably not going to be an option for me on the server.

Thanks for your responses. Solid looks very well designed, it's just whether i can buckle down enough to use it as my first js framework or not. :)

**Reply by ryansolid** [reply]
Caching is complicated to do right and easy to get unwanted stale data. One could cache stuff in the browser it just isn't always a default because it often isn't what you want. Ie.. value fresh data etc.. So most general demos won't do that. In the source I just have a global object that acts as a cache but I could picture just set/get from localStorage instead.

I mean invalidating it on something other than time is another thing. By the time you know it has changed it has already changed so you could be showing the latest. You could use Stale While Revalidate techniques but I find the experience jarring. Actually had a great conversation with Dan Abramov on the React core team about this a little while back.  This Twitter thread is full of good stuff: https://twitter.com/dan_abramov/status/1435117626926276608

**Reply by ITwrx** [reply]
> Caching is complicated to do right and easy to get unwanted stale data. One could cache stuff in the browser it just isn't always a default because it often isn't what you want. Ie.. value fresh data etc.. So most general demos won't do that.

Most assuredly! This is why i was wondering if there was some convenient, high level interface in Solid for developers to use (preferably almost blindly) that i just didn't know about yet. lol. Thanks for the info and link. I'll check that out. I can do Twiiter in ROBL (read-only by link) mode. :)

I'll keep an eye out for future Solid caching developments/docs. Feel free to ping me if you put something out.

Thanks

**Reply by ryansolid** [reply]
Yeah these sort of problems unless related to routing are probably outside of the core. There are some ecosystem libraries out there in frameworks (although not in Solid yet specifically), but like the thread says so little one size fits all. I find for the most part the data service fetching libraries take care of that.  If any really good solutions show up I can update this.

**Reply by ITwrx** [reply]
[Edited to correct my misunderstanding] 
Somehow notifying the client side from the front end that something has changed in the DB, was what i had in mind. Maybe SSE, IDK.  

No stale-while-revalidate. Just always have the latest data in the cache, and almost always serve from cache.

> I find for the most part the data service fetching libraries take care of that.

It wouldn't hurt my feelings if you had a couple suggestions on those. I don't know the ecosystem very well, but i understand if you're busy.

Thanks
